rpc/test/test.cpp

```diff
@@ -485,6 +485,123 @@ TEST_F(RpcTest, passive_shutdown) {
+        };
+    };
+    int do_rpc_service(Operation::Request* req, Operation::Response* resp, IOVector* iov, IStream* stream) {
+        resp->code = req->code;
+        return 0;
+    }
+    int do_rpc_service(OperationT::Request* req, OperationT::Response* resp, IOVector* iov, IStream* stream) {
+        resp->code = req->code;
+        photon::thread_usleep(5UL*1000*1000);
+        return 0;
+    }
+
+    int serve(photon::net::ISocketStream* stream) {
+        return m_skeleton->serve(stream);
+    }
+    int run() {
+        if (m_socket->bind_v4localhost() != 0)
+        // if (m_socket->bind(9527, net::IPAddr::V6Any()) != 0)
+            LOG_ERRNO_RETURN(0, -1, "bind failed");
+        if (m_socket->listen() != 0)
+            LOG_ERRNO_RETURN(0, -1, "listen failed");
+        m_endpoint = m_socket->getsockname();
+        LOG_DEBUG("bound to ", m_endpoint);
+        return m_socket->start_loop(false);
+    }
+    net::ISocketServer* m_socket;
+    Skeleton* m_skeleton;
+    photon::net::EndPoint m_endpoint;
+};
+
+static uint64_t do_call_hb(Stub* stub) {
+    RpcServerTimeout::Operation::Request req;
+    RpcServerTimeout::Operation::Response resp;
+    stub->call<RpcServerTimeout::Operation>(req, resp);
+    return 0;
+}
+
+TEST_F(RpcTest, timeout_with_hb) {
+    auto socket_server = photon::net::new_tcp_socket_server();
+    GTEST_ASSERT_NE(nullptr, socket_server);
+    DEFER(delete socket_server);
+    auto sk = photon::rpc::new_skeleton();
+    GTEST_ASSERT_NE(nullptr, sk);
+    DEFER(delete sk);
+
+    RpcServerTimeout rpc_server(sk, socket_server);
+    GTEST_ASSERT_EQ(0, rpc_server.run());
+
+    // photon::net::EndPoint ep(net::IPAddr::V4Loopback(), 9527);
+    auto& ep = rpc_server.m_endpoint;
+    auto pool = photon::rpc::new_stub_pool(-1, -1, 1'000'000);
+    DEFER(delete pool);
+    auto th1 = photon::thread_enable_join(photon::thread_create11([&]{
+        // Should always succeed in 3 seconds
+        auto stub = pool->get_stub(ep, false);
+        if (!stub) abort();
+        DEFER(pool->put_stub(ep, false));
+        Timeout timeout(7'000'000);
+        while(timeout.expired() > photon::now) {
+            int ret = do_call_hb(stub);
+            if (ret < 0) {
+                LOG_ERROR(VALUE(ret));
+                abort();
+            }
+            photon::thread_yield();
+        }
+    }));
+    photon::thread_yield_to((photon::thread*)th1);
+
+    auto th2 = photon::thread_enable_join(photon::thread_create11([&]{
+        // Should get connection refused after 2 seconds. Because socket closed listen fd at 1 second.
+        auto stub = pool->get_stub(ep, false);
+        if (!stub) {
+            abort();
+        }
+        DEFER(pool->put_stub(ep, false));
+        RpcServerTimeout::OperationT::Request req;
+        RpcServerTimeout::OperationT::Response resp;
+        auto before = photon::now;
+        auto ret = stub->call<RpcServerTimeout::OperationT>(req, resp, 1'000'000);
```

Comment:
If transfer timeout is 1s and rpc_call's timeout is 0.5s, you will find the end-to-end RPC cost time is still around 1s, which is very strange.

Reply:
> 4. `Stub::call` measures the time from invoking `call` before sending request to received
> response head. Receiving response body is not considered.

once got response head for the request, the new implementation will keep receiving body till success or another transfer timeout.

Since RPC framework using IStream as data stream, there is no API to set timeout during RPC receiving. Currently socket timeout can only set by ISockStream.

A better idea is add timeout for IStream, or make it possible set timeout for a single read/write operation for IStream. That will leads basic interface api change (IStream is the basic interface for IFile, ISocketStream, etc.), will need more discussion.

Reply:
Since new timeout methods are add to IStream, now transfer-timeout become no meaning.
RPC will controls timeout for a whole roundtrip.